### PR TITLE
[Chuong] function insert and get data from db + request permission

### PR DESCRIPTION
**Pull request by onstonboy**
function insert and get data from db + request permission

### CODE REVIEW

File: app/src/main/java/com/fstyle/androidtrainning/application/MainApplication.java [hunk 1]
```diff
@@ -12,10 +13,14 @@
 public class MainApplication extends Application {
 
     private static LastFmApi mApi;
+    private static Database mDatabase;
 
     @Override
     public void onCreate() {
         super.onCreate();
+        if (mDatabase == null){
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/application/MainApplication.java]: ERROR: '{' is not preceded with whitespace.

File: app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package com.fstyle.androidtrainning.data.local.roomdb.dao;
+
+import android.arch.persistence.room.Dao;
+import android.arch.persistence.room.Insert;
+import android.arch.persistence.room.Query;
+import com.fstyle.androidtrainning.data.local.roomdb.entity.TrackEntity;
+import com.fstyle.androidtrainning.model.Track;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java]: ERROR: Unused import: com.fstyle.androidtrainning.model.Track. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/mysong/subfragment/listsong/ListSongAdapter.java [hunk 3]
```diff
@@ -56,19 +71,63 @@ public int getItemCount() {
...
+
+        private void initViews(OnFavoriteClick onFavoriteClick) {
+            mImageView = itemView.findViewById(R.id.image_song);
             mTxtNameSong = itemView.findViewById(R.id.text_name_song);
             mTxtNameSinger = itemView.findViewById(R.id.text_name_singer);
+            mLikeButton = itemView.findViewById(R.id.image_favorite);
+            mOnFavoriteClick = onFavoriteClick;
+        }
+
+        private void handleEvents(final int pos) {
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/mysong/subfragment/listsong/ListSongAdapter.java]: WARNING: 'pos' hides a field.

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentFragment.java [hunk 4]
```diff
@@ -13,11 +13,13 @@
 import com.fstyle.androidtrainning.data.remote.service.config.LastFmApi;
 import com.fstyle.androidtrainning.model.RecentTracks;
 import com.fstyle.androidtrainning.screen.BaseFragment;
+import com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentFragment.java]: ERROR: Unused import: com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong.OnFavoriteClick. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java [hunk 5]
```diff
@@ -2,36 +2,29 @@
 
 import android.content.Context;
 import android.support.v7.widget.RecyclerView;
+import android.util.Log;
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java]: ERROR: Unused import: android.util.Log. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java [hunk 6]
```diff
@@ -2,36 +2,29 @@
...
 import android.view.ViewGroup;
 import android.widget.ImageView;
 import android.widget.TextView;
-
 import com.bumptech.glide.Glide;
 import com.fstyle.androidtrainning.R;
 import com.fstyle.androidtrainning.model.RecentTrack;
 import com.fstyle.androidtrainning.model.RecentTracks;
-
+import com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java]: ERROR: Unused import: com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong.OnFavoriteClick. Use Ctrl+Shift+O (Eclipse) or Ctrl+Alt+O (Intellij) to sort imports

File: app/src/main/java/com/fstyle/androidtrainning/screen/splash/SplashActivity.java [hunk 7]
```diff
@@ -47,6 +54,35 @@ public void run() {
...
+            doGoToMainActivity();
+        }
+    }
+
+    @Override
+    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
+            @NonNull int[] grantResults) {
+        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
+        switch (requestCode) {
+            case REQUEST_PERMISSION_CODE: {
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/splash/SplashActivity.java]: WARNING: Avoid nested blocks.

File: app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java (hunk 2 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java]: WARNING: [UnusedImports] Avoid unused imports such as 'com.fstyle.androidtrainning.model.Track'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentFragment.java (hunk 4 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentFragment.java]: WARNING: [UnusedImports] Avoid unused imports such as 'com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong.OnFavoriteClick'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java (hunk 5 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java]: WARNING: [UnusedImports] Avoid unused imports such as 'android.util.Log'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java (hunk 6 above)
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java]: WARNING: [UnusedImports] Avoid unused imports such as 'com.fstyle.androidtrainning.screen.main.mainfragments.mysong.subfragment.listsong.OnFavoriteClick'
https://pmd.github.io/pmd-5.6.1/pmd-java/rules/java/imports.html#UnusedImports

File: app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java [hunk 8]
```diff
@@ -131,18 +136,50 @@ private void setImagesRecent(int position) {
...
         public TracksViewHolder(View itemView) {
             super(itemView);
             initViews(itemView);
         }
 
+        private void initViews(View itemView) {
+            mImageView = itemView.findViewById(R.id.image);
+            mTxtSong = itemView.findViewById(R.id.text_upper);
+            mTxtSinger = itemView.findViewById(R.id.text_lower);
+        }
```
Comment (daolq3012) [inline app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java:174]: ERROR: [PMD-CPD] 22 lines duplicated.
app/src/main/java/com/fstyle/androidtrainning/screen/main/mainfragments/online/subfragment/recent/RecentRecyclerAdapter.java

File: app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+import java.util.List;
+
+/**
+ * Created by Administrator on 12/08/17.
+ */
+
+@Dao
+public interface TrackDAO {
+
+    @Insert
```
Comment (tranminhtri911) [inline app/src/main/java/com/fstyle/androidtrainning/data/local/roomdb/dao/TrackDAO.java]: @Insert(onConflict = OnConflictStrategy.REPLACE)

File: app/src/main/res/drawable-v21/ic_favorite.xml [hunk 10]
```diff
@@ -0,0 +1,4 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<vector xmlns:android="http://schemas.android.com/apk/res/android" android:height="24dp" android:viewportHeight="24.0" android:viewportWidth="24.0" android:width="24dp">
+    <path android:fillColor="#E53935" android:pathData="M12,21.35l-1.45,-1.32C5.4,15.36 2,12.28 2,8.5 2,5.42 4.42,3 7.5,3c1.74,0 3.41,0.81 4.5,2.09C13.09,3.81 14.76,3 16.5,3 19.58,3 22,5.42 22,8.5c0,3.78 -3.4,6.86 -8.55,11.54L12,21.35z"/>
```
Comment (tranminhtri911) [inline app/src/main/res/drawable-v21/ic_favorite.xml:3]: reformat code

**Comment by tranminhtri911** [comment]
Ok

**Pull request merged**